src/msgpack.lua

```diff
@@ -0,0 +1,220 @@
+local bit = require 'bit'
+local ffi = require 'ffi'
+local native = ffi.C
+local msgpack = require 'msgpack'
+
+local uint16_ptr_t = ffi.typeof('uint16_t *')
+local uint32_ptr_t = ffi.typeof('uint32_t *')
+
+ffi.cdef([[
+
+    enum mp_type {
+        MP_NIL = 0,
+        MP_UINT,
+        MP_INT,
+        MP_STR,
+        MP_BIN,
+        MP_ARRAY,
+        MP_MAP,
+        MP_BOOL,
+        MP_FLOAT,
+        MP_DOUBLE,
+        MP_EXT
+    };
+
+    union intbuff {
+        uint16_t u16;
+        uint32_t u32;
+        uint64_t u64;
+    };
+
+]])
+
+local char_ptr_t = ffi.typeof('char *')
+local bufp = ffi.new('const unsigned char *[1]');
+
+local function init_type_hints_array(format)
+    local xlat_table = {}
+    for _, value in ipairs(format) do
+        local from, to, filler = unpack(value)
+        for i=from,to,1 do
+            xlat_table[i] = filler
+        end
+    end
+    return xlat_table
+end
+
+local mp_type_hints = init_type_hints_array({
+    {0x00, 0x7f, native.MP_UINT}, -- MP_UINT (fixed)
+    {0x80, 0x8f, native.MP_MAP}, -- MP_MAP (fixed)
+    {0x90, 0x9f, native.MP_ARRAY}, -- MP_ARRAY (fixed)
+    {0xa0, 0xbf, native.MP_STR}, -- MP_STR (fixed)
+    {0xc0, 0xc0, native.MP_NIL}, -- MP_NIL
+    {0xc1, 0xc1, native.MP_EXT}, -- MP_EXT -- never used
+    {0xc2, 0xc3, native.MP_BOOL}, -- MP_BOOL
+    {0xc4, 0xc6, native.MP_BIN}, -- MP_BIN(8|16|32)
+    {0xc7, 0xc9, native.MP_EXT}, -- MP_EXT
+    {0xca, 0xca, native.MP_FLOAT}, -- MP_FLOAT
+    {0xcb, 0xcb, native.MP_DOUBLE}, -- MP_DOUBLE
+    {0xcc, 0xcf, native.MP_UINT}, -- MP_UINT
+    {0xd0, 0xd3, native.MP_INT}, -- MP_INT (8,16,32,64)
+    {0xd4, 0xd8, native.MP_EXT}, -- MP_INT? (8,16,32,64,127)
+    {0xd9, 0xdb, native.MP_STR}, -- MP_STR (8,16,32)
+    {0xdc, 0xdd, native.MP_ARRAY}, -- MP_ARRAY (16,32)
+    {0xde, 0xdf, native.MP_MAP}, -- MP_MAP (16,32)
+    {0xe0, 0xff, native.MP_INT}, -- MP_INT
+})
+
+local function mp_typeof(c)
```

Comment:
Not sure but maybe `is_mp_array`/`is_mp_map` helpers are enough for your task

Reply:
Well, I did not introduce these 1-time function helpers, but rather have reduced hints array to include only relevant to array and map information. Thus it's becoming quite obvious for which cases which hints would be returned. I assume observability has been improved, and code is still simple enough